### ISSUE TITLE
Bug: [no-unused-private-class-members] false positive with nested classes

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint#16117]
### Environment

Latest (website)

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
module.exports = {
  rules: {
    'no-unused-private-class-members': ['error']
  }
}
```
</details>

```js
class C {
    #usedInOuterClass;
    foo() {
        const fooThis = this;
        return class {
            #usedInOuterClass;
            bar() {
                return fooThis.#usedInOuterClass;
            }
        };
    }
}
```

[playground](https://eslint.org/play#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC1wcml2YXRlLWNsYXNzLW1lbWJlcnM6IFsnZXJyb3InXSAqL1xuXG5cbmNsYXNzIEMge1xuICAgICN1c2VkSW5PdXRlckNsYXNzO1xuICAgIGZvbygpIHtcbiAgICAgICAgY29uc3QgZm9vVGhpcyA9IHRoaXM7XG4gICAgICAgIHJldHVybiBjbGFzcyB7XG4gICAgICAgICAgICAjdXNlZEluT3V0ZXJDbGFzcztcbiAgICAgICAgICAgIGJhcigpIHtcbiAgICAgICAgICAgICAgICByZXR1cm4gZm9vVGhpcy4jdXNlZEluT3V0ZXJDbGFzcztcbiAgICAgICAgICAgIH1cbiAgICAgICAgfTtcbiAgICB9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

### What did you expect to happen?

Error on the inner class, or no error at all.

### What actually happened?

An error on the outer class:

```
Error:
5:5 '#usedInOuterClass' is defined but never used. (no-unused-private-class-members)
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

This is getting into really weird code territory - but I figured it was worth documenting this limitation.

The rule is designed to consume the class contexts from inside to outside:
https://github.com/eslint/eslint/blob/49ca3f090425e06fdf6e66bcf2415508c46671e1/lib/rules/no-unused-private-class-members.js#L78
https://github.com/eslint/eslint/blob/49ca3f090425e06fdf6e66bcf2415508c46671e1/lib/rules/no-unused-private-class-members.js#L97
And it assumes that there aren't any name clashes.

It's sadly impossible for this to be correctly handled by the rule because there's no type information to correctly understand which class instance is being referenced.

I would propose the following:
1) inspect the context in which the identifier is used. If it's a `this.#priv` or `#priv in this` - then we know almost for certain it must be referencing the top of the stack.
    - it's illegal to have code like `function () { this.#priv }.bind(fooThis)`, so we know there's no possible rebinding weirdness.
2) in other cases - assume *any* class in the stack is being referenced, and mark **all** of them as used.

For (2) you could technically go one step further and use scope analysis attempt to resolve what the variable refers to. This would let you handle the error case as described above - i.e. would reduce the false-negative rate. Though TBH it's probably not worth going to such effort to handle code that is already super bad.

**Comment by bradzacher** [comment]
Ah damn - whilst this case is syntactically valid - I now realise in testing that it's not semantically correct, and would error at runtime because you're accessing a private property on `Foo` within `Bar`.

So there's no need to worry about the false-positive here.

**Comment by mdjermanovic** [comment]
> 2\. in other cases - assume _any_ class in the stack is being referenced

`#foo` always refers to `#foo` of the innermost enclosing class that defines a `#foo` private member. Private member basically shadows private members that have the same name ([9.2.1.2 ResolvePrivateIdentifier](https://tc39.es/ecma262/#sec-resolve-private-identifier)).

> ```js
> class C {
>     #usedInOuterClass;
>     foo() {
>         const fooThis = this;
>         return class {
>             #usedInOuterClass;
>             bar() {
>                 return fooThis.#usedInOuterClass;
>             }
>         };
>     }
> }
> ```

This is not a false positive on the outer `#usedInOuterClass`, because `#usedInOuterClass` in `fooThis.#usedInOuterClass` does not refer to it. Now, this could be seen as a false negative on the inner `#usedInOuterClass`, because it is used only on an object that doesn't really have that private member, but that would be difficult to determine without types, and it does look more like a type-checking error than a lint error.
